MagneticField/GeomBuilder/src/DD4hep_volumeHandle.cc

```diff
@@ -197,9 +203,483 @@ namespace {
+  double rIn = convertUnits(tubs.rMin());               // inner radius
+  double rOut = convertUnits(tubs.rMax());              // outer radius
+  double startPhi = tubs.startPhi();                    // angular start of the tube-section
+  double deltaPhi = tubs.deltaPhi();                    // angular span of the tube-section
+  double cutAtStart = convertUnits(tubs.cutAtStart());  // truncation at begin of the tube-section
+  double cutAtDelta = convertUnits(tubs.cutAtDelta());  // truncation at end of the tube-section
+  bool cutInside = tubs.cutInside();                    // true, if truncation is on the inner side of the tube-section
+
+  LogTrace("MagGeoBuilder") << "zhalf      " << zhalf << newln << "rIn        " << rIn << newln << "rOut       " << rOut
+                            << newln << "startPhi   " << startPhi << newln << "deltaPhi   " << deltaPhi << newln
+                            << "cutAtStart " << cutAtStart << newln << "cutAtDelta " << cutAtDelta << newln
+                            << "cutInside  " << cutInside;
+
+  //
+  //  GlobalVector planeXAxis = refPlane->toGlobal(LocalVector( 1, 0, 0));
+  //  GlobalVector planeYAxis = refPlane->toGlobal(LocalVector( 0, 1, 0));
+  GlobalVector planeZAxis = refPlane->toGlobal(LocalVector(0, 0, 1));
+
+  Sides cyl_side;
+  Sides plane_side;
+  double rCyl = 0;
+  double rCentr = 0;
+  if (cutInside) {
+    cyl_side = outer;
+    rCyl = rOut;
+    plane_side = inner;
+    rCentr = (max(max(rIn, cutAtStart), cutAtDelta) + rOut) / 2.;
+  } else {
+    cyl_side = inner;
+    rCyl = rIn;
+    plane_side = outer;
+    rCentr = (rIn + min(min(rOut, cutAtStart), cutAtDelta)) / 2.;
+  }
+
+  // Recalculate center: (for a DDTruncTubs, DDD gives 0,0,Z)
+  // The R of center is in the middle of the arc of cylinder fully contained
+  // in the trunctubs.
+  Geom::Phi<double> phiCenter(startPhi + deltaPhi / 2.);
+  center_ = refPlane->toGlobal(LocalPoint(rCentr * cos(phiCenter), rCentr * sin(phiCenter), 0.));
+
+  // FIXME!! Actually should recompute RN from pos_Rplane; should not
+  // matter anyhow
+  theRN = rCentr;
+
+  // For simplicity, the position of the cut plane is taken at one of the
+  // phi edges, where R is known
+  // FIXME! move to center of plane
+  // FIXME: compute. in double prec (not float)
+  GlobalPoint pos_Rplane(refPlane->toGlobal(LocalPoint(LocalPoint::Cylindrical(cutAtStart, startPhi, 0.))));
+
+  // Compute angle of the cut plane
+  // Got any easier formula?
+  // FIXME: check that it still holds for cutAtStart > cutAtDelta
+  double c = sqrt(cutAtDelta * cutAtDelta + cutAtStart * cutAtStart - 2 * cutAtDelta * cutAtStart * cos(deltaPhi));
+  double alpha = startPhi - asin(sin(deltaPhi) * cutAtDelta / c);
+  GlobalVector x_Rplane = refPlane->toGlobal(LocalVector(cos(alpha), sin(alpha), 0));
+  Surface::RotationType rot_R(planeZAxis, x_Rplane);
+
+  // FIXME: use builder
+  surfaces[plane_side] = new Plane(pos_Rplane, rot_R);
```

Comment:
for my understanding, where are these objects deleted later? Going through the volumeHandle base classes, I was not immediately able to idenfity it, but I wasn't looking in the right place probably.

Reply:
Surfaces are reference counted. cf. https://github.com/cms-sw/cmssw/blob/master/DataFormats/GeometrySurface/interface/Surface.h#L36